src/core/internal/hash.d

```diff
@@ -230,6 +388,22 @@ size_t hashOf(T)(auto ref T val, size_t seed = 0) if (!is(T == enum) && (is(T ==
             return bytesHashAlignedBy!T(bytes, seed);
         }
     }
+};
+
+//struct or union hash
+size_t hashOf(T)(scope const auto ref T val, size_t seed = 0)
+if (!is(T == enum) && (is(T == struct) || is(T == union))
+    && canBitwiseHash!T)
+{
+    mixin(_hashOfStruct);
+}
+
+//struct or union hash
+size_t hashOf(T)(auto ref T val, size_t seed = 0)
+if (!is(T == enum) && (is(T == struct) || is(T == union))
+    && !canBitwiseHash!T)
+{
+    mixin(_hashOfStruct);
```

Comment:
Why do we need the duplication here?

Reply:
One of the reasons that `canBitwiseHash!T` could be false for a struct is that it has a callable `toHash`, so we don't really know that without checking.

Reply:
For instance:
https://run.dlang.io/is/q43nKg
```d
void main()
{
    enum hasCallableToHash(T) = __traits(compiles, {
            size_t hash = ((T* x) => (*x).toHash())(null);
        });

    static struct TaggedObject
    {
        Object _object;
        size_t tag;
        alias _object this;
    }

    static assert(hasCallableToHash!TaggedObject); // Passes.
    static assert(hasCallableToHash!(const(TaggedObject))); // Fails!
}
```

Reply:
Oh I was more questioning the `scope const` duplication, which should be valid anyhow (see e.g. https://run.dlang.io/is/qpKrds for your example).
But yeah unfortunately we can't assume that `toHash` is `const` :/

Reply:
I don't think the compiler actually checks anything related to `scope` or I misunderstand what `scope` means on a function argument. https://run.dlang.io/is/JEg0hH